providers/hns/hns_roce_u_verbs.c

```diff
@@ -442,67 +483,124 @@ static int hns_roce_alloc_srq_buf(struct hns_roce_srq *srq)
 	return 0;
 }
 
-struct ibv_srq *hns_roce_u_create_srq(struct ibv_pd *pd,
-				      struct ibv_srq_init_attr *init_attr)
+static int hns_roce_verify_srq(struct hns_roce_context *context,
+			       struct ibv_srq_init_attr_ex *init_attr)
 {
-	struct hns_roce_context *ctx = to_hr_ctx(pd->context);
-	struct hns_roce_create_srq_resp resp;
-	struct hns_roce_create_srq cmd;
+	if (init_attr->srq_type != IBV_SRQT_BASIC &&
+	    init_attr->srq_type != IBV_SRQT_XRC)
+		return -EINVAL;
+
+	if (!init_attr->attr.max_wr || !init_attr->attr.max_sge ||
+	    init_attr->attr.max_wr > context->max_srq_wr ||
+	    init_attr->attr.max_sge > context->max_srq_sge)
+		return -EINVAL;
+
+	return 0;
+}
+
+static int exec_srq_create_cmd(struct ibv_context *context,
+			       struct hns_roce_srq *srq,
+			       struct ibv_srq_init_attr_ex *init_attr)
+{
+	struct hns_roce_create_srq_resp resp = {};
+	struct hns_roce_create_srq_ex cmd_ex = {};
+	struct hns_roce_ib_create_srq *cmd_drv;
+	struct hns_roce_create_srq cmd = {};
+	bool is_basic_srq;
+	int ret;
+
+	is_basic_srq = (init_attr->srq_type == IBV_SRQT_BASIC) ||
+		       (!(init_attr->comp_mask & IBV_SRQ_INIT_ATTR_TYPE));
+
+	if (is_basic_srq)
+		cmd_drv = &cmd.drv_payload;
+	else
+		cmd_drv = &cmd_ex.drv_payload;
+
+	cmd_drv->buf_addr = (uintptr_t)srq->buf.buf;
+	cmd_drv->que_addr = (uintptr_t)srq->idx_que.buf.buf;
+	cmd_drv->db_addr = (uintptr_t)srq->db;
+
+	if (is_basic_srq)
```

Comment:
This if isn't needed always use cerate_srq_ex

Reply:
Thank you, I will fix it